### ISSUE TITLE
Debug Adapter crashes while debugging when printing JSON in [brackets]

### ISSUE THREAD

**Issue by mit-mit** [Dart-Code/Dart-Code#483]
Repro steps:

Create new app.

Replace main.dart with:
```
import 'package:flutter/material.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      home: new Scaffold(
        appBar: new AppBar(
          title: new Text('repro'),
        ),
        body: new Center(
          child: new RaisedButton(
            child: new Text('foo'),
            onPressed: () {
              print(
                  '[{"author":"foo@example.com"},{"company":"Example Inc."},{"creationDate":1513934676},{"description":""},{"title":""},{"type":3},{"typeDescription":"Other"},{"uploaded":0}]');
            },
          ),
        ),
      ),
    );
  }
}
```

Hit F5

Press the button in the app.

Dart Code crashes:
![screen shot 2017-12-22 at 10 34 28](https://user-images.githubusercontent.com/13644170/34293117-b6a0ebc0-e703-11e7-98de-f7297f997d26.png)

**Comment by DanTup** [comment]
Ok, I think what's happening here is that `print` just outputs directly to stdout and when we use `flutter run` we read from stdout in order to get commands back. We only parse anything that's wrapped in `[brackets]` as commands, but since your output is, we assume it's a command.

I can put something in to stop the crash in this case (eg. check it's both valid json and also has an `id` property), however it wouldn't be difficult to break things again by sending some real JSON that does have an ID and it would be interpreted as a command.

@devoncarew Does this problem exist in other IDEs? Is there a good way around it? I think ideally when running `flutter run --machine` the users `prints` wouldn't come raw down `stdout`?

**Comment by DanTup** [comment]
Currently it seems we don't even do a json check (which would fix this particular instance):

https://github.com/Dart-Code/Dart-Code/blob/4633e60e3d7493a01daddfe23e1b949206743bfc/src/debug/flutter_run.ts#L18-L24

@mit-mit as a temporary workaround, if you ensure your message doesn't start with `[` and end with `]` it'll stop the crash.

**Comment by mit-mit** [comment]
Ah, interesting. I'm not blocked on this, just hit it as I was doing some print line debugging (simply printing out a json object I saving to disc).

**Comment by DanTup** [comment]
Ok, I've made some tweaks:

Invalid JSON will no longer silently be dropped but instead assumed to be user-output. Previously if you just did `print("[test]");` the output wouldn't appear in your console at all (it would appear in the debug adapters debug output, but you wouldn't see that unless you were debugging the debug adapter).

Valid JSON that does not have an `event` or `id` property will now also be assumed to be user-output and also be printed to the console (as with above, it'll output a warning if debugging the debug adapter about this assumption)...

Users debug console:

<img width="749" alt="screen shot 2017-12-25 at 19 58 48" src="https://user-images.githubusercontent.com/1078012/34342678-091327f8-e9af-11e7-9cc7-f2f89ec2f74c.png">

Debug console if debugging the debug adapter:

<img width="1149" alt="screen shot 2017-12-25 at 19 58 56" src="https://user-images.githubusercontent.com/1078012/34342679-092e65b8-e9af-11e7-9565-c2da488e1ca9.png">

I believe this will solve most of these issues, however there is still a case that if a user `prints` something that is valid JSON, inside [square brackets] that has an `id` or `event` property, bad things may still happen. I'd like to chat with @devoncarew about this at some point - the ideal IMO would be to wrap users stdout but I have limited knowledge of this so don't know if it's feasible or even a good idea.

**Comment by DanTup** [comment]
Oops, bad copy/paste in the second screenshot there - this one should not say `non-` in front of JSON. Fixed now, though that'll only ever be seen by anyone debugging the debug adapter (most likely just me and motivated contributors!).

**Comment by DanTup** [comment]
I was just trying to test another issue and wanted to reproduce this, but couldn't.

When I checked the log, the output from the print statement above looks like this:

```text
[15:15:48]: <== I/flutter (13637): [{"author":"foo@example.com"},{"company":"Example Inc."},{"creationDate":1513934676},{"description":""},{"title":""},{"type":3},{"typeDescription":"Other"},{"uploaded":0}]
```

Note it starts with "I/flutter" which means my code didn't try to parse it, and therefore didn't crash. The only difference I can think is that it was iOS when I repro'd this but Android here. @devoncarew Any idea why this might be different? Is it possible iOS is missing some info that would've avoided this, or is the prefix from Android unexpected/unwanted?

**Comment by devoncarew** [comment]
Hey, the solution above looks reasonable to me - more vetting of the `[{...}]` output. We can definitely look at a more out-of-band solution for delivering user text, though this one does have the benefit of being simple.

iOS and Android do log `print()` slightly differently. iOS will have no prefix, and android will have the prefix we get from adb (which qualifies where the output came from a bit).

**Comment by DanTup** [comment]
The problem with the current fix is that if the user sends something that looks like a flutter command (eg. has an id) it'll still cause issues. Although it's an edge case, it's kinda weird if a user raises a bug for "If I log JSON that has an id field, I get this error. Probably we'd then add more validation/filtering, but feels like a game of cat & mouse!

That said, my latest comment was more about the difference - is it expected (desired) to be prefixed for Android and not iOS? Although it does the opposite of fix this, I think pre-fixing in Android is kinda weird - if the user prints some text out, they'd expected to see it raw. Is this just a side-effect of something, or a conscious decision?

**Comment by Skylled** [comment]
I'm pretty sure all output from adb is prefixed with the type of output (a letter, I for Info, E for Error, etc) and where it came from (a word, like Flutter), and that's just how adb has been.

**Comment by DanTup** [comment]
Yeah, I thought that might be the case; though I don't think it'd be crazy for Flutter to strip it off (or send stdout over a different channel back to the client); I think the tooling in the middle should probably be transparent to the user. When I do `print("test")` I'd only really expected to see `test` in the output.

**Comment by devoncarew** [comment]
It's a side effect of how we get stdout from iOS vs android. For android, we rely on adb logging to get stdout.

That's a decision from when we originally built the tooling - we had no other option at the time. We now have a consistent debug connection to the device, so we could choose to do something like:
- not display any adb output in flutter run
- have `flutter run` listen to the debug connection for `Stdout` and `Stderr` events (the VM fires events to the service protocol for write to stdout and stderr)

Something to consider. It would mean that you would not see stdout from apps unless you have a debug connection hooked up and are logging it.

**Comment by devoncarew** [comment]
We could probably do all the work in flutter_tools, so no changes necessary in the framework, engine.

**Comment by DanTup** [comment]
Ok; well it's not causing my significant issues (it's more with my user hat on I think it's weird) so I'll leave it to you to decide if you think it warrants any work (if you want me to raise an issue at flutter, lmk).

**Comment by DanTup** [comment]
This [came up again](https://github.com/Dart-Code/Dart-Code/issues/494#issuecomment-357426197) - despite my changes above a user managed to crash by printing `[null]`. This looked like a flutter command (square brackets) so I `JSON.parse`'d the contents then checked if it had an `event` property but it crashed being null!

I've made more changes to improve this, but we are just chipping away at the edge cases so I definitely think there's merit in a better fix. It probably won't be long before someone prints some JSON that looks even more like a flutter command and this happens again :(

I'll put a case in to try/catch around this handling so it won't result in a debugger crash, but it still might result in something strange.
